source/agora/utils/TxBuilder.d

```diff
@@ -105,42 +128,102 @@ public struct TxBuilder
 
     public this (in PublicKey refundMe) @safe pure nothrow
     {
-        this.leftover = Output(Amount(0), refundMe);
-        this.unlocker = &TxBuilder.keyUnlocker;
+        this.builder = TransactionBuilder(&KeyUnlocker, refundMe);
     }
 
     /// Ditto
     public this (in Lock lock) @safe pure nothrow
     {
-        this.leftover = Output(Amount(0), lock);
-        this.unlocker = &TxBuilder.keyUnlocker;
+        this.builder = TransactionBuilder(&KeyUnlocker, lock);
     }
 
     /// Ditto
     public this (const Transaction tx) @safe nothrow
     {
-        this(tx.outputs[0].lock);
-        this.attach(tx);
+        this.builder = TransactionBuilder(&KeyUnlocker, tx);
     }
 
     /// Ditto
     public this (const Transaction tx, uint index) @safe nothrow
     {
-        this(tx.outputs[index].lock);
+        this.builder = TransactionBuilder(&KeyUnlocker, tx, index);
+    }
+
+    /// Ditto
+    public this (const Transaction tx, uint index, in Lock lock)
+        @safe nothrow
+    {
+        this.builder = TransactionBuilder(&KeyUnlocker, tx, index, lock);
+    }
+
+    /// Convenience constructor that calls `this.attach(Output, Hash)`
+    public this (in Output utxo, in Hash hash) @safe nothrow
+    {
+        this.builder = TransactionBuilder(&KeyUnlocker, utxo, hash);
+    }
+}
+
+///
+public struct TransactionBuilder
```

Comment:
to me, this feels like it was better suited to be an abstract class. we wouldn't need all the ctor wrappers in `StaticTransactionBuilder`, in fact we wouldnt need `StaticTransactionBuilder`

Reply:
Yes I think you are right. Will have a look at updating it. If it starts to take too long then will come back to it later.